internal/bus/kafka.go

```diff
@@ -0,0 +1,186 @@
+		return nil, fmt.Errorf("unable to configure sasl mechanism (%s)", saslMechanismName)
+	}
+}
+
+func Initialize() error {
+	var tlsConfig *tls.Config
+	var saslMechanism sasl.Mechanism
+
+	// configure TLS when CA certificate was provided
+	if config.Kafka.CACert != "" {
+		cleanCAPath := filepath.Clean(config.Kafka.CACert)
+		pemCerts, err := ioutil.ReadFile(cleanCAPath)
+		if err != nil {
+			return fmt.Errorf("unable to read kafka CA cert: %w", err)
+		}
+
+		pool := x509.NewCertPool()
+		pool.AppendCertsFromPEM(pemCerts)
+
+		tlsConfig = &tls.Config{
+			MinVersion: tls.VersionTLS13,
+			RootCAs:    pool,
+		}
+	}
+
+	// configure SASL if mechanism was provided
+	if config.Kafka.SASL.SaslMechanism != "" {
+		var err error
+		saslMechanism, err = createSaslMechanism(config.Kafka.SASL.SaslMechanism, config.Kafka.SASL.Username, config.Kafka.SASL.Password)
+		if err != nil {
+			return fmt.Errorf("kafka SASL error: %w", err)
+		}
+	}
+
+	dialer = &kafka.Dialer{
+		ClientID:      version.KafkaClientID,
+		Timeout:       10 * time.Second,
+		DualStack:     true,
+		SASLMechanism: saslMechanism,
+		TLS:           tlsConfig,
+	}
+
+	transport = &kafka.Transport{
+		Dial:     dialer.DialFunc,
+		ClientID: version.KafkaClientID,
+		TLS:      tlsConfig,
+		SASL:     saslMechanism,
+	}
+
+	return nil
+}
+
+func newContextLogger(ctx context.Context) func(msg string, a ...interface{}) {
+	return func(msg string, a ...interface{}) {
+		logger := ctxval.Logger(ctx)
+		logger.Debug().Bool("kafka", true).Msgf(msg, a...)
+	}
+}
+
+func newContextErrLogger(ctx context.Context) func(msg string, a ...interface{}) {
+	return func(msg string, a ...interface{}) {
+		logger := ctxval.Logger(ctx)
+		logger.Warn().Bool("kafka", true).Msgf(msg, a...)
+	}
+}
+
+func NewReader(ctx context.Context, topic string) *kafka.Reader {
+	return kafka.NewReader(kafka.ReaderConfig{
+		Brokers:     config.Kafka.Brokers,
+		Dialer:      dialer,
+		Topic:       topic,
+		Logger:      kafka.LoggerFunc(newContextLogger(ctx)),
+		ErrorLogger: kafka.LoggerFunc(newContextErrLogger(ctx)),
+	})
+}
+
+// NewWriter creates synchronous writer created from the pool. Make sure to close
+// it with Close() function.
+func NewWriter(ctx context.Context, topic string) *kafka.Writer {
+	return &kafka.Writer{
+		Addr:        kafka.TCP(config.Kafka.Brokers...),
+		Topic:       topic,
+		Transport:   transport,
+		Logger:      kafka.LoggerFunc(newContextLogger(ctx)),
+		ErrorLogger: kafka.LoggerFunc(newContextErrLogger(ctx)),
+	}
+}
+
+// BatchConsume reads messages in batches up to 1 MB with up to 10 seconds delay. Use
+// context cancellation to stop the loop.
+func BatchConsume(ctx context.Context, topic string, handler func(message *GenericMessage)) {
+	logger := ctxval.Logger(ctx)
+	r := NewReader(ctx, topic)
+	defer r.Close()
+
+	for {
+		kmsg, err := r.ReadMessage(ctx)
+		if err != nil && errors.Is(err, io.EOF) {
+			break
+		} else {
+			logger.Warn().Err(err).Msg("error when reading message")
```

Comment:
This logs an error even if `err == nil`, right?

Reply:
Not sure what you mean, but the intention was:

* If there was an error, but it was end of file, then break the loop (and exit the goroutine).
* If there was any other error, just print a warning in log and continue reading.
* Otherwise log a trace log message and deliver the message into the callback handler function.

Do you see there a problem? I have to be missing something.

Reply:
Yes, you fixed this with the [6e5197c](https://github.com/RHEnVision/provisioning-backend/commit/6e5197c3c82f38595811089a2127a01a0a84ca1b) push. :+1: Not relevant anymore.

Reply:
Ah rebasing is bad, I should stop doing that and push separate commits. Its just I am worried about us accidentally merging all of them without squash :-)